marker_lints/src/not_using_has_span_trait.rs

```diff
@@ -0,0 +1,44 @@
+use marker_api::prelude::*;
+
+marker_api::declare_lint! {
+    /// ### What it does
+    /// Suggests using `impl HasSpan` for functions that need to take `Span` as
+    /// a parameter to make them more ergonomic.
+    NOT_USING_HAS_SPAN_TRAIT,
+    Warn,
+}
+
+pub(crate) fn check_item<'ast>(cx: &'ast MarkerContext<'ast>, item: ItemKind<'ast>) {
+    let ItemKind::Fn(func) = item else { return };
+
+    for param in func.params() {
+        let ast::TyKind::Path(path) = strip_references(param.ty()) else {
+            continue;
+        };
+
+        let ident = path
+            .path()
+            .segments()
+            .last()
+            .unwrap_or_else(|| panic!("BUG: path must have at least one segment, but got: {path:#?}"))
+            .ident();
+
+        if ident != "Span" {
+            continue;
+        }
```

Comment:
I didn't find a way to resolve the semantic information of an `ast::TyKind`. That's why I had to use a heuristic that compares the last segment of the path with `Span` which is good enough, but definitely not ideal in cases when there are multiple crates with multiple types named `Span`.

Reply:
Jep, you're right. There should be an option to check this using [`resolve_ty_ids`](https://docs.rs/marker_api/latest/marker_api/context/struct.MarkerContext.html#method.resolve_ty_ids), but there seems to be no way of getting the `TyDefId`, woops :sweat_smile:

Reply:
The problem with `resolve_ty_ids` is that it takes a stringified path as a parameter. It also returns a list of matching types.
In this case we should have a method that takes an `ast::TyKind` as a parameter and returns just a single `sem::TyKind`. I suppose it shouldn't be ambiguous, and only a single type can be resolved given the context of the crate and module where the `ast::TyKind` resides it should always resolve to a single type, right?

Reply:
Rustc has a util to convert AST types to semantic types, it can crash under some circumstances, which is why I'm cautious with it. I think the better option would be to have a `resolve_item_ids` equivalent. The AST path of the `ast::TyKind` should already provide the `ItemId` it refers to. Right now, there is just no way to check if that's the one from our `Span` struct